### ISSUE TITLE
Performance Analysis workshop for PSycloneBench

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSycloneBench#97]
I figured I'd make an issue to detail results with different tools we're using as part of the PA workshop.

Probably we'll look at both NemoLite (incl distributed memory) and tracer advection.

First tool is Vtune, and i'll post (and update) things as I get results.
@vkatkinson If you also have results to share it would be good to place here as well probably.

**Comment by LonelyCat124** [comment]
Tracer Advection + vtune:
This seems to primarily be memory bound - particularly with cross NUMA accesses being a significant amount of memory accesses. I'll show results here with both
```
export OMP_PLACES=cores
export OMP_PROC_BIND=close/spread
```

**Comment by sergisiso** [comment]
Which version of psyclone are use using: master or latest release? What script do you use for generation?

I thought in master we generate no schedule clause by default instead of auto (that we did for a short period).
To change it you can use:  `omp_loop_trans.omp_schedule = "none"` or "static". By none we mean that we don't want a clause at all (which is implementation defined and different compilers do different things depending on architecture)

**Comment by LonelyCat124** [comment]
I'm on master (downloaded yesterday), with the default `make tra_adv_omp_cpu` script. 

From looking in PSyclone, the `OMPLoopTrans` has this init flag:
`def __init__(self, omp_directive="do", omp_schedule="auto"):`
so I guess the default here is `auto`, even though for the directive its set to `none`.

**Comment by LonelyCat124** [comment]
NemoLite2D has a very different characteristic as far as I can tell.

I tried to remove a lot of the serial writes and things, but there still appears to be a lot present.

![image](https://user-images.githubusercontent.com/3043914/235175847-8c8e97ef-57ac-4075-9371-7d6bf49e43d0.png)
![image](https://user-images.githubusercontent.com/3043914/235175939-a75f928a-e418-4fe7-ad1f-17e2f02426c4.png)


I need to dive down a bit more at some point into this, but:
1. OpenMP imbalance could speedup the OpenMP sections by ~8% - which doesn't seem too bad.
2. Presumably NemoLite2D arrays are not initialised with OpenMP, so we have terrible NUMA behaviour (which we can probably mitigate with 2 MPI procs per rank).
3. Much less memory bound (I assume higher arithmetic intensity)
4. Almost none of NemoLite2D vectorises, I'll try to explore why, but at a guess I assume its just too many if statements to be worthwhile vectorising.

**Comment by LonelyCat124** [comment]
Ok - I setup MAQAO on Scafellpike, and testing it with NemoLite2D at the moment.

scripts to run:
lsf file
```
#BSUB -J test
#BSUB -e test.err
#BSUB -o test.out
#BSUB -x
#BSUB -n <number_nodes>
#BSUB -R "span[ptile=<number_processes_per_node>]"
#BSUB -W 01:00

export OMP_NUM_THREADS=32
export OMP_PLACES=cores
export OMP_PROC_BIND=close
<mpi_command> <run_command>
```

MAQAO lua file (minus comments):
```
experiment_name = nil
comments = nil
executable     = "./nemolite2d.exe"
external_libraries = {}
run_command    = "<executable>"
batch_script    = "maqao.lsf"
batch_command   = "bsub < <batch_script>"
number_processes = 1
number_nodes = 1
number_processes_per_node = 1
mpi_command    = "mpirun -n <number_processes>"


filter = {
   type = "number",
   value = 15,
}

-- !! Uncomment the table you want to use !!
-- If the profiling should begin after a given percentage of the application time.
-- A first run of the application is automatically performed to time the application.
--profile_start = {
--   unit = "p",      -- 'p' for 'percentage'
--   value = 30,         -- delay in percentage of the total application time
--}
-- If the profiling should begin when the application is started
--profile_start = {
--   unit = "none",   -- Specify that no delay is needed
--}

-- If the profiling should be triggered from the application using source probes
-- Use help command $maqao oneview --help-guided-profile
-- to learn how to add probes to the source code
--profile_start = {
--   unit = "probe",  -- Uses LPROF probes
--}

-- If the profiling should begin after a given time (in second)
--profile_start = {
--   unit = "s",      -- 's' for 'seconds'
--   value = 0,         -- delay in seconds
--}

dataset        = ""
dataset_handler= "link"
```

The results seem to match pretty reasonably with VTune's suggestions - vectorisation is low:
![image](https://user-images.githubusercontent.com/3043914/236199549-2033ee69-5a74-4f72-8507-f8a251ecb899.png)

![image](https://user-images.githubusercontent.com/3043914/236199692-6f2f299b-7ec7-4204-ada3-d834f686e4b3.png)

Given this is essentially due to boundary condition checks in loops it seems hard to remove for now. Short of that, performance is pretty reasonable.

**Comment by LonelyCat124** [comment]
Also ran a scalability run with it and NemoLite2D
r0 i don't know what ran - error on my side i assume
r1 = 1 rank 32 OMP
r2 = 32 rank 1 OMP
r3 = 2 rank 16 OMP
r4 = 2 rank 32 OMP - 2nodes
r5 = 4 rank 16 OMP - 2 nodes

![image](https://user-images.githubusercontent.com/3043914/236235192-8217c1df-cb1c-41c4-8087-686bdd64ba67.png)


Nemolite clearly scales much better with MPI only than with OpenMP.

**Comment by arporter** [comment]
> This comes from not telling the compiler that these arrays are `contiguous`

Is this an artefact of the way we access the arrays from within a field object in NEMOLite2D? Now that I come to think about it, we did quite an in-depth look at the performance of NEMOLite2D in a previous paper. I'll see if I can find it.

**Comment by arporter** [comment]
Also, which version of the OMP-enabled build are you running with (as I think there are different options)?

**Comment by LonelyCat124** [comment]
> Is this an artefact of the way we access the arrays from within a field object in NEMOLite2D? Now that I come to think about it, we did quite an in-depth look at the performance of NEMOLite2D in a previous paper. I'll see if I can find it.

Yes I believe so, though I think it might just be any time you access an array inside a function since the compiler in fortran doesn't know if its been passed in as `fullarray` or `array_section(1:3, 4:5)` for example.

I think I'm just doing `make nemolite2d_omp` and `make nemolite2d_hybrid`

**Comment by arporter** [comment]
That paper his here: https://gmd.copernicus.org/articles/11/3447/2018/gmd-11-3447-2018.pdf
